### PR TITLE
Validation Script for OC Path and RPC Coverage in READMEs

### PR DESCRIPTION
**Pull request by wenovus**
* `tools/internal/mdocspec/*` package parses ocpath.proto and ocrpcs.proto from yaml according to the format shown below by mdocspec.Parse's documentation.
* `tools/validate_readme_spec/validate_readme_spec.go` script is added that can parse and then validate OC Paths and RPCs according to existing repo states in github.com/openconfig.

```go
// Parse extracts sorted OpenConfig Path and RPC Coverage from an FNT README.
//
// The first yaml code block after a heading line named exactly as
// "OpenConfig Path and RPC Coverage" will be used. Any other code blocks are
// ignored.
//
// Expected markdown format:
//
//      ## OpenConfig Path and RPC Coverage
//
//      ```yaml
//      paths:
//        /interfaces/interface/config/description:
//        /interfaces/interface/config/enabled:
//        /components/component/state/name:
//          platform_type: "CHASSIS"
//
//      rpcs:
//        gnmi:
//          gNMI.Set:
//            union_replace: true
//          gNMI.Subscribe:
//            on_change: true
//      ```
//
// Note: For `rpcs`, only the RPC name and methods are validated. Any
// attributes defined below RPC methods (e.g. union_replace) are not validated.
func Parse(source []byte) (*ppb.OCPaths, *rpb.OCRPCs, error) {
```

### CODE REVIEW

**Comment by wenovus** [comment]
This validation routine is based on the template README style being defined in https://github.com/openconfig/featureprofiles/pull/2863

File: .github/workflows/readme_oc_path_and_rpc.yml
Comment (dplore) [inline .github/workflows/readme_oc_path_and_rpc.yml:36]: With a syntax change in it's yaml section, this readme could be added: 

https://github.com/openconfig/featureprofiles/blob/main/feature/bgp/policybase/otg_tests/import_export_multi/README.md

Comment (dplore) [reply .github/workflows/readme_oc_path_and_rpc.yml:36]: In a future PR, we could change the default to parse README with changes perhaps?  Or introduce a flag as an exception in metadata.textproto for all old README's that haven't had a yaml stanza added.

File: tools/internal/mdocspec/md_test.go
Comment (dplore) [inline tools/internal/mdocspec/md_test.go:265]: I think I would like to see the language required, just for pretty rendering for humans to read.  WDYT?

File: tools/internal/mdocspec/md_test.go
Comment (dplore) [inline tools/internal/mdocspec/md_test.go:12]: should one of the tests take it's input from the test-requirements-template.md ?    This would help remind us to update this check when we change the template in the future?

File: tools/internal/mdocspec/ocspec.go
Comment (dplore) [inline tools/internal/mdocspec/ocspec.go:21]:
```suggestion
// Parse extracts sorted OpenConfig Path and RPC Coverage from a featureprofiles README.
```

File: tools/internal/mdocspec/ocspec.go
Comment (dplore) [inline tools/internal/mdocspec/ocspec.go:19]: OpenConfig Path and RPC Coverage should probably instead refer to the constant `OCSpecHeading`

Maybe https://github.com/openconfig/featureprofiles/blob/main/doc/test-requirements-template.md should be in a constant as well?

File: tools/internal/mdocspec/md.go
Comment (dplore) [inline tools/internal/mdocspec/md.go:19]: Is this the depth of the tree to search?

File: tools/internal/mdocspec/md_test.go
Comment (wenovus) [reply tools/internal/mdocspec/md_test.go:265]: This is the intended behaviour, improved error desc string to make clearer.

File: .github/workflows/readme_oc_path_and_rpc.yml
Comment (wenovus) [reply .github/workflows/readme_oc_path_and_rpc.yml:36]: I will make the change in a follow-up PR to validate the file.

I propose to parse all READMEs all the time, since that's the only way to know the repo is in good state all the time.

I think the best way to identify READMEs is to parse all `README.md` files under `/features` repo, which is what `addrundata` also does. `addrundata` also validates that whenever a `_test.go` file exists, `README.md` must also exist, which helps our check as well.

In fact, I propose only to tolerate `ErrNotFound` any README that does not have an accompanying `_test.go` file. This way we won't need allow-lists either.

File: tools/internal/mdocspec/md_test.go
Comment (wenovus) [reply tools/internal/mdocspec/md_test.go:12]: I already have an integration tests taking in `test-requirements-template.md` as part of CI.

File: tools/internal/mdocspec/ocspec.go
Comment (wenovus) [reply tools/internal/mdocspec/ocspec.go:19]: Done.

The template file is harder to make a constant. I think we should strive to not change it.

File: tools/internal/mdocspec/md.go
Comment (wenovus) [reply tools/internal/mdocspec/md.go:19]: The goldmark parser I'm using creates a hierarchy of MD nodes, and in general there can be several sibling nodes in between the OCSpecHeading and the actual yaml block. This specifies the number of previous sibling nodes to search for OCSpecHeading before giving up.

File: .github/workflows/readme_oc_path_and_rpc.yml
Comment (dplore) [reply .github/workflows/readme_oc_path_and_rpc.yml:33]: I would like to cover README yaml path validation even when the README is standalone / no _test.go file.

File: tools/internal/mdocspec/md.go
Comment (dplore) [reply tools/internal/mdocspec/md.go:19]: Hrm, sorry I still don't follow.   In any case, I expect the yaml block to appear immediately after / within the OCSpecHeading.  There could be 10+  headings in a big or highly parameterized test.  So maybe make this number 20?

Comment (wenovus) [reply tools/internal/mdocspec/md.go]: Thanks for the suggestion, I changed the logic to search from the heading rather than from the YAML block. This way we can just search until the next heading of the same or lower level or the end of the document, which is cleaner since we can remove this constant.

File: .github/workflows/readme_oc_path_and_rpc.yml
Comment (wenovus) [reply .github/workflows/readme_oc_path_and_rpc.yml:33]: I'm still proposing to validate all README.md under /features **where the heading exists**, no matter whether a _test.go file is present. However, I'm proposing that if a _test.go file doesn't exist, we don't force the issue. The benefit of this is that we don't need an allow-list.

WDYT? -- an allow-list or don't force the test author to write such a section if a _test.go file doesn't exist?

Comment (dplore) [reply .github/workflows/readme_oc_path_and_rpc.yml:33]: I feel like we should require the section to exist without exceptions.    (It requires converting existing README before merging these checks in).  Worst case an author can add an empty section with a TODO comment?

Comment (wenovus) [reply .github/workflows/readme_oc_path_and_rpc.yml:33]: Alright, allow-list approach implemented here: https://github.com/openconfig/featureprofiles/pull/2881

Comment (dplore) [reply .github/workflows/readme_oc_path_and_rpc.yml:33]: thanks!

File: tools/internal/mdocspec/md.go
Comment (dplore) [inline tools/internal/mdocspec/md.go:83]: Should we continue parsing and throw an error if there is more than one YAML block?  ie: there must be exactly one YAML block in the ocSpecHeading?   I am not sure we really care if there are other yaml blocks in other headings.

File: tools/internal/mdocspec/ocspec.go
Comment (dplore) [inline tools/internal/mdocspec/ocspec.go:126]: nit, but in test format, we use want before got?

```suggestion
		return nil, nil, fmt.Errorf("mdocspec: Want at least one RPC method in the YAML, but got 0.")
```

File: tools/internal/ocpaths/clone_public.go
Comment (dplore) [inline tools/internal/ocpaths/clone_public.go:22]: could this cause detection of an old/stale repo?

File: tools/internal/ocrpcs/ocrpcs.go
Comment (dplore) [inline tools/internal/ocrpcs/ocrpcs.go:24]: same comment as above re: stale repo copies

File: tools/internal/mdocspec/md.go
Comment (wenovus) [reply tools/internal/mdocspec/md.go:83]: I think it's ok to ignore YAML blocks aside from the first one. I don't think it's realistic to achieve the goal of making sure that the author doesn't put multiple coverage-related blocks into the same README. Even if they did, the duplicate block should be caught in review anyways, so I don't think this will be an actual issue. WDYT?

File: tools/internal/mdocspec/ocspec.go
Comment (wenovus) [reply tools/internal/mdocspec/ocspec.go:126]: In Go tests we generally have got before want: https://google.github.io/styleguide/go/decisions.html#got-before-want. However since this is not a test I didn't consider this point. Is there some other value in changing the wording here?

File: tools/internal/ocpaths/clone_public.go
Comment (wenovus) [reply tools/internal/ocpaths/clone_public.go:22]: Yes, the purpose is to avoid having to download every single time. Let me add more documentation on this, thanks for catching.

File: tools/internal/ocrpcs/ocrpcs.go
Comment (wenovus) [reply tools/internal/ocrpcs/ocrpcs.go:24]: Same purpose as above. Go doc comments updated.

File: tools/internal/mdocspec/md.go
Comment (dplore) [reply tools/internal/mdocspec/md.go:83]: Sure, good enough as is

File: tools/internal/mdocspec/ocspec.go
Comment (dplore) [reply tools/internal/mdocspec/ocspec.go:126]: Updated wording works for me!

**Review by dplore — DISMISSED** [review]

**Comment by wenovus** [comment]
@dplore Need another approval due to slight doc fix.

**Pull request merged**
